generator_nodemodules.go

```diff
@@ -0,0 +1,95 @@
+package dalec
+
+import (
+	"path/filepath"
+
+	"github.com/moby/buildkit/client/llb"
+	"github.com/pkg/errors"
+)
+
+func (s *Source) isNodeMod() bool {
+	for _, gen := range s.Generate {
+		if gen.NodeMod != nil {
+			return true
+		}
+	}
+	return false
+}
+
+// HasNodeMods returns true if any of the sources in the spec are node modules.
+func (s *Spec) HasNodeMods() bool {
+	for _, src := range s.Sources {
+		if src.isNodeMod() {
+			return true
+		}
+	}
+	return false
+}
+
+func withNodeMod(g *SourceGenerator, srcSt, worker llb.State, opts ...llb.ConstraintsOpt) func(llb.State) llb.State {
+	return func(in llb.State) llb.State {
+		workDir := "/work/src"
+		joinedWorkDir := filepath.Join(workDir, g.Subpath)
+		srcMount := llb.AddMount(workDir, srcSt)
+		installCmd := "npm install"
+		// Run install in a new state, mounting the source
+		installed := worker.Run(
+			ShArgs(installCmd),
+			llb.Dir(joinedWorkDir),
+			srcMount,
+			WithConstraints(opts...),
+		).AddMount(filepath.Join(joinedWorkDir, "node_modules"), in)
```

Comment:
What is "in" in this case?

I feel like we should be able to `.AddMount(joinedWorkDir, srcSt)` and avoid the copy operation below.